File: backend/src/main/java/com/onlinestore/backend/controller/ProductController.java
package com.onlinestore.backend.controller;

import com.onlinestore.backend.model.Product;
import com.onlinestore.backend.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.util.*;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    @Autowired
    private ProductRepository productRepository;

    private static final String UPLOAD_DIR = "uploads/";

    @GetMapping
    public ResponseEntity<?> getProducts(
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) String sort) {

        Sort sortObj = Sort.by(Sort.Direction.DESC, "createdAt");
        if ("price_asc".equals(sort)) {
            sortObj = Sort.by(Sort.Direction.ASC, "price");
        } else if ("price_desc".equals(sort)) {
            sortObj = Sort.by(Sort.Direction.DESC, "price");
        }

        List<Product> products;
        
        if (search != null && !search.isEmpty()) {
            products = productRepository.searchProducts(search, sortObj);
        } else if (category != null && !category.isEmpty()) {
            products = productRepository.findByCategory(category, sortObj);
        } else {
            products = productRepository.findAll(sortObj);
        }

        Map<String, Object> response = new HashMap<>();
        response.put("count", products.size());
        response.put("products", products);

        return ResponseEntity.ok(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getProductById(@PathVariable String id) {
        Optional<Product> productOpt;
        if (id.startsWith("PRD-")) {
            productOpt = productRepository.findByProductId(id);
        } else {
            productOpt = productRepository.findById(id);
        }

        if (productOpt.isPresent()) {
            Product product = productOpt.get();
            product.setViews(product.getViews() + 1);
            productRepository.save(product);
            return ResponseEntity.ok(product);
        } else {
            return ResponseEntity.status(404).body(Map.of("message", "Product not found"));
        }
    }

    @PostMapping
    public ResponseEntity<?> createProduct(
            @RequestParam("name") String name,
            @RequestParam("category") String category,
            @RequestParam("price") Double price,
            @RequestParam("stockQuantity") Integer stockQuantity,
            @RequestParam("description") String description,
            @RequestParam(value = "material", required = false) String material,
            @RequestParam(value = "size", required = false) String size,
            @RequestParam(value = "files", required = false) MultipartFile[] files) {

        List<String> imagePaths = new ArrayList<>();
        if (files != null) {
            for (MultipartFile file : files) {
                try {
                    Files.createDirectories(Paths.get(UPLOAD_DIR));
                    String fileName = UUID.randomUUID().toString() + "_" + file.getOriginalFilename();
                    Path path = Paths.get(UPLOAD_DIR + fileName);
                    Files.write(path, file.getBytes());
                    imagePaths.add("/uploads/" + fileName);
                } catch (IOException e) {
                    return ResponseEntity.status(500).body(Map.of("message", "File upload failed"));
                }
            }
        }

        String generatedId = "PRD-" + java.time.LocalDate.now().toString().replace("-", "") + "-" +
                UUID.randomUUID().toString().substring(0, 4).toUpperCase();

        Product product = new Product();
        product.setProductId(generatedId);
        product.setName(name);
        product.setCategory(category);
        product.setPrice(price);
        product.setStockQuantity(stockQuantity);
        product.setDescription(description);
        product.setMaterial(material);
        product.setSize(size);
        product.setImages(imagePaths);
        product.setCreatedAt(LocalDateTime.now());
        product.setUpdatedAt(LocalDateTime.now());

        Product savedProduct = productRepository.save(product);
        return ResponseEntity.status(201).body(savedProduct);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateProduct(
            @PathVariable String id,
            @RequestParam(value = "name", required = false) String name,
            @RequestParam(value = "category", required = false) String category,
            @RequestParam(value = "price", required = false) Double price,
            @RequestParam(value = "stockQuantity", required = false) Integer stockQuantity,
            @RequestParam(value = "description", required = false) String description,
            @RequestParam(value = "material", required = false) String material,
            @RequestParam(value = "size", required = false) String size,
            @RequestParam(value = "files", required = false) MultipartFile[] files) {

        Optional<Product> productOpt = productRepository.findById(id);

        if (productOpt.isPresent()) {
            Product product = productOpt.get();
            if (name != null) product.setName(name);
            if (category != null) product.setCategory(category);
            if (price != null) product.setPrice(price);
            if (stockQuantity != null) product.setStockQuantity(stockQuantity);
            if (description != null) product.setDescription(description);
            if (material != null) product.setMaterial(material);
            if (size != null) product.setSize(size);

            if (files != null && files.length > 0) {
                List<String> newImages = new ArrayList<>();
                for (MultipartFile file : files) {
                    try {
                        Files.createDirectories(Paths.get(UPLOAD_DIR));
                        String fileName = UUID.randomUUID().toString() + "_" + file.getOriginalFilename();
                        Path path = Paths.get(UPLOAD_DIR + fileName);
                        Files.write(path, file.getBytes());
                        newImages.add("/uploads/" + fileName);
                    } catch (IOException e) {
                        return ResponseEntity.status(500).body(Map.of("message", "File upload failed"));
                    }
                }
                if (product.getImages() == null) {
                    product.setImages(newImages);
                } else {
                    product.getImages().addAll(newImages);
                }
            }

            product.setUpdatedAt(LocalDateTime.now());
            Product updatedProduct = productRepository.save(product);
            return ResponseEntity.ok(updatedProduct);
        } else {
            return ResponseEntity.status(404).body(Map.of("message", "Product not found"));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteProduct(@PathVariable String id) {
        if (productRepository.existsById(id)) {
            productRepository.deleteById(id);
            return ResponseEntity.ok(Map.of("message", "Product removed"));
        } else {
            return ResponseEntity.status(404).body(Map.of("message", "Product not found"));
        }
    }

    @GetMapping("/recommendations/{category}")
    public ResponseEntity<?> getRecommendations(@PathVariable String category, @RequestParam(required = false) String exclude) {
        List<Product> products;
        if (exclude != null) {
            products = productRepository.findByCategoryAndIdNot(category, exclude);
        } else {
            products = productRepository.findByCategory(category, Sort.by(Sort.Direction.DESC, "createdAt"));
        }
        
        if (products.size() > 4) {
            products = products.subList(0, 4);
        }
        return ResponseEntity.ok(products);
    }

    @GetMapping("/trending")
    public ResponseEntity<?> getTrendingProducts() {
        List<Product> products = productRepository.findAll(Sort.by(Sort.Direction.DESC, "views"));
        if (products.size() > 4) {
            products = products.subList(0, 4);
        }
        return ResponseEntity.ok(products);
    }

    @GetMapping("/recent")
    public ResponseEntity<?> getRecentProducts() {
        List<Product> products = productRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"));
        if (products.size() > 4) {
            products = products.subList(0, 4);
        }
        return ResponseEntity.ok(products);
    }
}

File: backend-old/controllers/productController.js
const Product = require('../models/Product');

// @desc    Fetch all products
// @route   GET /api/products
// @access  Public
const getProducts = async (req, res) => {
    try {
        const query = {};
        
        // Search functionality
        if (req.query.search) {
            query.$or = [
                { name: { $regex: req.query.search, $options: 'i' } },
                { productId: { $regex: req.query.search, $options: 'i' } }
            ];
        }

        // Filter by category
        if (req.query.category) {
            query.category = req.query.category;
        }

        // Sorting
        let sortObj = { createdAt: -1 }; // default: newest first
        if (req.query.sort === 'price_asc') {
            sortObj = { price: 1 };
        } else if (req.query.sort === 'price_desc') {
            sortObj = { price: -1 };
        }

        const products = await Product.find(query).sort(sortObj);
        
        res.json({
            count: products.length,
            products
        });
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Fetch single product by ID or custom productId
// @route   GET /api/products/:id
// @access  Public
const getProductById = async (req, res) => {
    try {
        let product;
        if (req.params.id.startsWith('PRD-')) {
            product = await Product.findOne({ productId: req.params.id });
        } else {
            product = await Product.findById(req.params.id);
        }

        if (product) {
            // Increment views for trending feature
            product.views = (product.views || 0) + 1;
            await product.save();
            
            res.json(product);
        } else {
            res.status(404);
            throw new Error('Product not found');
        }
    } catch (error) {
        res.status(404).json({ message: 'Product not found' });
    }
};

// @desc    Create a product
// @route   POST /api/products
// @access  Private/Admin
const createProduct = async (req, res) => {
    try {
        const { name, category, price, stockQuantity, description, material, size } = req.body;
        
        // Handle images if uploaded
        let images = [];
        if (req.files && req.files.length > 0) {
            images = req.files.map(file => `/uploads/${file.filename}`);
        }

        // Generate Custom Product ID
        const dateStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
        const randomStr = Math.random().toString(36).substring(2, 6).toUpperCase();
        const generatedId = `PRD-${dateStr}-${randomStr}`;

        const product = new Product({
            productId: generatedId,
            name,
            category,
            price,
            stockQuantity,
            description,
            material,
            size,
            images
        });

        const createdProduct = await product.save();
        res.status(201).json(createdProduct);
    } catch (error) {
        res.status(400).json({ message: error.message });
    }
};

// @desc    Update a product
// @route   PUT /api/products/:id
// @access  Private/Admin
const updateProduct = async (req, res) => {
    try {
        const { name, category, price, stockQuantity, description, material, size } = req.body;

        const product = await Product.findById(req.params.id);

        if (product) {
            product.name = name || product.name;
            product.category = category || product.category;
            product.price = price !== undefined ? price : product.price;
            product.stockQuantity = stockQuantity !== undefined ? stockQuantity : product.stockQuantity;
            product.description = description || product.description;
            product.material = material || product.material;
            product.size = size || product.size;

            // Handle new images upload (append or replace depending on logic)
            // For now, let's append new images to existing ones.
            if (req.files && req.files.length > 0) {
                const newImages = req.files.map(file => `/uploads/${file.filename}`);
                product.images = [...product.images, ...newImages];
            }

            const updatedProduct = await product.save();
            res.json(updatedProduct);
        } else {
            res.status(404);
            throw new Error('Product not found');
        }
    } catch (error) {
        res.status(400).json({ message: error.message });
    }
};

// @desc    Delete a product
// @route   DELETE /api/products/:id
// @access  Private/Admin
const deleteProduct = async (req, res) => {
    try {
        const product = await Product.findById(req.params.id);

        if (product) {
            await Product.deleteOne({ _id: product._id });
            res.json({ message: 'Product removed' });
        } else {
            res.status(404);
            throw new Error('Product not found');
        }
    } catch (error) {
        res.status(404).json({ message: error.message });
    }
};

// --- FEATURE 1: RECOMMENDATIONS ---

// @desc    Get similar products based on category
// @route   GET /api/products/recommendations/:category
// @access  Public
const getRecommendations = async (req, res) => {
    try {
        const { category } = req.params;
        // Optionally exclude current product by passing ?exclude=id
        const excludeId = req.query.exclude;
        
        const query = { category: category };
        if (excludeId) {
            query._id = { $ne: excludeId };
        }

        // Limit to 4 for the UI
        const products = await Product.find(query).limit(4);
        res.json(products);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Get trending products based on views or stock movement
// @route   GET /api/products/trending
// @access  Public
const getTrendingProducts = async (req, res) => {
    try {
        // Sort by views descending, limit to 4
        const products = await Product.find({}).sort({ views: -1 }).limit(4);
        res.json(products);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Get recently added products
// @route   GET /api/products/recent
// @access  Public
const getRecentProducts = async (req, res) => {
    try {
        // Sort by createdAt descending, limit to 4
        const products = await Product.find({}).sort({ createdAt: -1 }).limit(4);
        res.json(products);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

module.exports = {
    getProducts,
    getProductById,
    createProduct,
    updateProduct,
    deleteProduct,
    getRecommendations,
    getTrendingProducts,
    getRecentProducts
};
